### PR TITLE
Add tests for PET BMI module (replaces PR#401)

### PR DESCRIPTION
**Pull request by stcui007**
Add PET module and related .github test runners

## Additions

data/example_bmi_multi_realization_config_w_noah_pet_cfe__linux.json
data/example_bmi_multi_realization_config_w_noah_pet_cfe__macos.json

## Removals

-

## Changes

.github/workflows/module_integration.yml
.github/workflows/test_and_validate.yml
include/realizations/catchment/Bmi_Module_Formulation.hpp

## Testing

unit test
bmi tests
compare_pet

## Notes

-

## Todos

-

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [X] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [X] New functions are documented (with a description, list of inputs, and expected output)
- [X] Placeholder code is flagged / future todos are captured in comments
- [X] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [X] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

## Testing checklist (automated report can be put here)

1.

### Target Environment support

- [X] Linux

### CODE REVIEW

**Comment by mattw-nws** [comment]
So your Linux build problem happens at linking, and it looks like after #417 (specifically 891d6c2 ) you will need to include `{$NETCDF_LIBRARIES}` in the `compare_pet` target (see the `compare_cfe` target right before it in CMakeLists.txt in `test` directory), or set the `use_netcdf` parameter of the build action to `off`. This may also solve your macOS build problem but I haven't looked at it.

You also seem to be getting a linker issue with pthreads, but linking against `core::mediator` should include that, so I'm not sure why that's happening... solve the NetCDF problem first and see what happens.

**Comment by stcui007** [comment]
Thanks.
I was looking at the error details and was wondering how netCDF got
involved since Bmi_C_Pet_IT.cpp doesn't use netCDF.
Also Bmi_C_Pet_IT.cpp uses CsvPerFeatureForcingProvider so that probably
needs to be considered also. Not sure why compare_pet doesn't fail in
testing on UCS machines when using Forcing.h.

**Comment by mattw-nws** [comment]
> Not sure why compare_pet doesn't fail in testing on UCS machines when using Forcing.h

I don't think the use of `Forcing.h` (or `CsvPerFeatureForcingProvider.h` for that matter) in `Bmi_C_Pet_IT.cpp` file does anything. The forcing params are passed up to the `Formulation_Manager` which obtains the forcings via whatever the config says.

As for the code building successfully in other environments, that probably depends on the build options used. The point of my comment above is that #417 defaulted building with NetCDF to "on" in the Github workflows... if it's not on in your build, then it probably works fine. That said, if you turn NetCDF support on in your build, it will probably break like it is here in the workflow.

**Comment by stcui007** [comment]
Made some progress, able to link to netCDF. But still getting the  "Can't
init bmi_c_pet; unreadable shared library file ''" error.

File: data/example_bmi_multi_realization_config_w_noah_pet_cfe__linux.json [hunk 1]
```diff
@@ -0,0 +1,90 @@
+{
```
Comment (mattw-nws) [inline data/example_bmi_multi_realization_config_w_noah_pet_cfe__linux.json:1]: This file is a copy of the one without `__linux`...

**Review by mattw-nws — CHANGES_REQUESTED** [review]
Since there's a redundant/copy file in here and #430 got done at the same time, lets remove both the `__linux` and `__macos` realization files and edit the remaining realization config to remove the extension (see changes to JSON and YAML in #430).

**Comment by mattw-nws** [comment]
Okay those changes look good but you did a merge on Github (probably) that should have been a rebase: [acd4a49](https://github.com/NOAA-OWP/ngen/pull/429/commits/acd4a494592a29532ecbda9884eb41a288e4d233)

**Comment by stcui007** [comment]
I did a rebase to upstream master and then merged the master into the ngen_devPET.

**Comment by stcui007** [comment]
All were done on linux, not github.

**Comment by mattw-nws** [comment]
The merge step there is unnecessary... the rebase is the merge. I think if you rebase onto upstream/master again that merge commit will disappear. Try that and push (force) again.

**Comment by stcui007** [comment]
Do I run the rebase on the ngen_devPET branch directly or run it on the
master branch and then 'git checkout ngen_devPET'? Does force push
overwrite the changes I made to module_integration.yml and realization json?

**Comment by mattw-nws** [comment]
```
git checkout ngen_devPET
git rebase upstream/master # assumes you have a remote named upstream
git log # commit acd4a49 should be missing but everything else should be the same
git push --force ngen_devPET # will push what you just saw in git log
```

**Comment by stcui007** [comment]
I tried to run "git pull --rebase upstream master" in the ngen_devPET, it
failed because there are changes in module_integration.yml.

**Comment by stcui007** [comment]
I got an error:
$ git rebase upstream/master
fatal: Needed a single revision
invalid upstream upstream/master

$ git remote -v
origin  https://github.com/stcui007/ngen (fetch)
origin  https://github.com/stcui007/ngen (push)
upstream        https://github.com/NOAA-OWP/ngen.git (fetch)
upstream        https://github.com/NOAA-OWP/ngen.git (push)

**Comment by stcui007** [comment]
$git log
commit e6139ba76cded7c40249d4a12e043462e9630ee3
Author: Shengting Cui <shengting.cui@noaa.gov>
Date:   Thu Aug 4 18:42:49 2022 +0000

    Make changes to accomodate upstream master change using extensionless lib file name.

commit acd4a494592a29532ecbda9884eb41a288e4d233
Merge: 7f3e05e f64ffd3
Author: Shengting Cui <shengting.cui@noaa.gov>
Date:   Thu Aug 4 18:15:19 2022 +0000

    Merge branch 'master' into ngen_devPET

commit f64ffd336dbea7eaa1ad5848dd79ce59b35594a1
Author: Matt Williamson <matthew.williamson@gov>
Date:   Wed Aug 3 20:54:13 2022 +0000

    Added extensionless solution also.

**Review by mattw-nws — APPROVED** [review]

**Review by mattw-nws — APPROVED** [review]

**Pull request merged**